### PR TITLE
Implement parsing EraVM assembler

### PR DESCRIPTION
**Pull request by atrosinenko**
This PR contains implementation of assembler parser for EraVM target as well as several refactorings made to backend's internal representation of EraVM instructions.

Related issues: #411, #412, #414, #416, #417, #418, #419, #420, #421, #422, #423, #435.

### CODE REVIEW

**Comment by akiramenai** [comment]
It looks like there are plenty of functional and non-functional changes combined in this PR, so it's hard to review. Can we divide the functionality to meaningful commits and review them one by one? I mean divide to less than 114 commits; form a commit structure as we actually plan to land these changes.
Also please clang format and pass the linter.

**Comment by asl** [comment]
> It looks like there are plenty of functional and non-functional changes combined in this PR, so it's hard to review. Can we divide the functionality to meaningful commits and review them one by one? I mean divide to less than 114 commits; form a commit structure as we actually plan to land these changes.

The commits itself are quite well-divided and represent fine-grained incremental set of changes, I would say there are almost no non-functional changes. However, we had to restructure the instructions definitions entirely as in the original form they were unsuitable for instruction encoding and had some inherent issues. Certainly, some commits could be combined, for example:

 - Support parsing name.s for commutative instructions (such as add.s)
 - Support parsing "name.s reg, reg, reg" for non-commutative instructions
 - Parse context.<cop> instructions
 - ...

Let me know if you'd want us to do this.

**Comment by akiramenai** [comment]
@asl it sounds like a reasonable change. I think we should go for it.
Could we also make sure no diff `clang-format`, `clang-tidy` produce no warnings, and `check-llvm` pass for all commits?

**Comment by asl** [comment]
> Could we also make sure no diff `clang-format`, `clang-tidy` produce no warnings, and `check-llvm` pass for all commits?

We are working on silencing linter warnings.

**Comment by asl** [comment]
@hedgar2017 @akiramenai @sayon 

Looks like the integration tests fail because they use standalone assembler that cannot handle the assembler syntax described in the spec. How would you want to address this?

**Comment by akiramenai** [comment]
@asl 
It looks like we miss a register as the first argument of `ret.to_label` if to satisfy our assembler. It should be easy to fix.
But in general, if I remember it right, we agreed to stick with the current syntax and made the changes requested by @hedgar2017, @sayon in separate patches. Which looks like a sane approach. So, with that in mind, passing the integration tests shouldn't be that hard, and the sooner we integrate, the better. We become confident in our asm printer this way.

**Comment by asl** [comment]
@akiramenai Yes, the renaming is a part of https://github.com/matter-labs/compiler-llvm/issues/441

However, it looks like there are 3 different sources of information that do no correspond to each other: backend, assembler and the ISA spec. There were some obsolete instructions, some obsolete syntax, some undocumented syntax, etc. So, in such cases we had to resort to the existing spec as an ultimate source of information. Some changes (e.g. removal of obsolete stuff) is certainly a part of initial code drop as it does not make any sense to make them supported only to remove them immediately afterwards. 

In this particular case, it seems that standalone assembler does not support `revert` instruction alias as specified in ISA spec. We will add remained aliases to please the assembler, certainly :)

**Comment by akiramenai** [comment]
@asl 
I understand that it's our fault. We haven't tested the spec against the assembler. So thank you for finding the discrepancy. We can raise an asm issue there and fix it. The fix could likely be landed on Monday. If you can do an alias, that could unblock you faster.
I still think it worth to integrate earlier. When we pass system contract compilation, the CI won't stop after the first failure.

**Comment by akiramenai** [comment]
@asl I asked @sayon, [cc @shamatar]
But I believe the spec is wrong here. It doesn't mention the argument for `ret`, `revert` but to the best of my memory we pass the pointer to the returning slice, which makes sense if it's a far return or revert. But as we don't distinguish near and far in case of returns and reverts, we always pass an argument.
N.B. @sayon is OOO today.

**Comment by hedgar2017** [comment]
@akiramenai @asl @sayon I'd kindly ask you guys to remove all aliases as soon as possible (e.g. the `revert` mentioned above). If it's just for the initial phase to reach conformance, that's fine.
It would be great to start over without legacy.

**Comment by asl** [comment]
> @akiramenai @asl @sayon I'd kindly ask you guys to remove all aliases as soon as possible (e.g. the `revert` mentioned above). If it's just for the initial phase to reach conformance, that's fine.

Absolutely. However, here we are having chicken-and-egg problem: existing assembler supports only some aliases, but not others (e.g. `ret.revert` but not `revert`) or always require register for `ret.ok` regardless what is in the spec, even for default one, also it seems that it supports only some subset of specified assembler syntax (as it comes with stack operands, for example). So we'd need to add everything and then remove legacy :)

However, we structured the changes in such way so it will be more or less straightforward.

**Comment by atrosinenko** [comment]
I pushed a few commits that (temporary?) rollback potentially breaking changes to syntax. Let's see if tests pass.

**Comment by sayon** [comment]
I have pushed more updates to the the spec available online; now the returns are properly categorized:
- far return/revert and near return/revert are merged and became  `ret reg` and `rev reg`. Register argument is ignored if we return from near calls.
- returns/reverts/panics to label  are marked "near" because far returns ignore labels
- `pnc` (panic) does not accept arguments because panic ignores register argument

This type specifies the abstract syntax of the programmer-facing assembler instructions:
https://matter-labs.github.io/eravm-spec/spec.html#asm_instruction

**Comment by asl** [comment]
@akiramenai What does "exec format error" error mean?

**Comment by hedgar2017** [comment]
> > @akiramenai @asl @sayon I'd kindly ask you guys to remove all aliases as soon as possible (e.g. the `revert` mentioned above). If it's just for the initial phase to reach conformance, that's fine.
> 
> Absolutely. However, here we are having chicken-and-egg problem: existing assembler supports only some aliases, but not others (e.g. `ret.revert` but not `revert`) or always require register for `ret.ok` regardless what is in the spec, even for default one, also it seems that it supports only some subset of specified assembler syntax (as it comes with stack operands, for example). So we'd need to add everything and then remove legacy :)
> 
> However, we structured the changes in such way so it will be more or less straightforward.

Understood!
@akiramenai do you think we need to compare both assemblers for binary equality, or would a simple migration be sufficient? Assembly is low-level enough for our testset to walk all the paths, so I don't need to implement another comparison mode on compiler-tester, especially as a one-time job.

**Comment by akiramenai** [comment]
@asl, I have never seen that before. But from the fact that only Viper programs are affected I infer that it's likely to be a CI issue. Could you rebase the work on ToT?
@hedgar2017 could you also have a look at the integration tests failures. You should know it better.

**Comment by asl** [comment]
> @akiramenai do you think we need to compare both assemblers for binary equality, or would a simple migration be sufficient?

Here is an idea: we do have a set of assembler files (under `test/MC/EraVM`) to test the functionality. We can just run the existing assembler and`llvm-mc` and compare the outputs? Is there a quick way to dump the encoding from the existing assembler?

**Comment by akiramenai** [comment]
@hedgar2017 we can migrate I think. Binary comparison is faster though, could have less dependence, and could identify issues that aren't touched by input data.

**Comment by hedgar2017** [comment]
> @asl, I have never seen that before. But from the fact that only Viper programs are affected I infer that it's likely to be a CI issue. Could you rebase the work on ToT? @hedgar2017 could you also have a look at the integration tests failures. You should know it better.

I can only see LIT failures on CI. Could you point me directly?

**Comment by hedgar2017** [comment]
> @hedgar2017 we can migrate I think. Binary comparison is faster though, could have less dependence, and could identify issues that aren't touched by input data.

Binary comparison isn't faster than simply running the CI after migration. I'll see if I can do some hacking like dump all builds from compiler-tester.

**Comment by atrosinenko** [comment]
> I can only see LIT failures on CI. Could you point me directly?

@hedgar2017 Adjusted recently added `rol.ll` and `ror.ll` tests for the minor changes in SP-increment syntax, now `check-llvm` should pass.

**Comment by atrosinenko** [comment]
These "Exec format error" messages look like something returned by `exec()` syscall to me - maybe something related to `binfmt-misc` configuration or trying to execute binaries for incorrect CPU architecture.

**Comment by asl** [comment]
Now integration tests seem to pass. So likely was some random CI breakage. Please review

File: llvm/lib/Target/EraVM/AsmParser/EraVMAsmParser.cpp
Comment (akiramenai) [inline llvm/lib/Target/EraVM/AsmParser/EraVMAsmParser.cpp]: Could we avoid capturing everything by reference where it's redundant? We only use `Str`, and it can be passed by value. Same applies to other lambdas.

File: llvm/lib/Target/EraVM/EraVMOpcodes.td
Comment (sayon) [inline llvm/lib/Target/EraVM/EraVMOpcodes.td]: Please update the filename/description in comment.

File: llvm/lib/Target/EraVM/EraVMExpandPseudoInsts.cpp
Comment (sayon) [inline llvm/lib/Target/EraVM/EraVMExpandPseudoInsts.cpp]: IIUC this is a predicate; if so, would using `COND_NONE` instead of magic numbers be preferable everywhere in the codebase?

In the current codebase there are no instruction building facilities that would properly decompose an instruction structurally, or compose it out of the parts, therefore `addImm(0)` is used both as a part of code/stack composed operand, and as a predicate. It may look confusing.

File: llvm/lib/Target/EraVM/EraVMInstrFormats.td
Comment (akiramenai) [inline llvm/lib/Target/EraVM/EraVMInstrFormats.td]: tables?

File: llvm/lib/Target/EraVM/EraVMInstrFormats.td
Comment (akiramenai) [inline llvm/lib/Target/EraVM/EraVMInstrFormats.td:105]: I introduced that duplication because I had issues with `OperandAddrMode` as the ColField. As you refactor our `td`s anyway, may I ask if it's hard to get rid of `OperandAM`?

File: llvm/lib/Target/EraVM/EraVMInstrFormats.td
Comment (sayon) [inline llvm/lib/Target/EraVM/EraVMInstrFormats.td:183]: Thank you for separating the encoding from the logical instruction set!

File: llvm/lib/Target/EraVM/EraVMInstrInfo.td
Comment (akiramenai) [inline llvm/lib/Target/EraVM/EraVMInstrInfo.td]: disassembler?

File: llvm/lib/Target/EraVM/EraVMInstrFormats.td
Comment (sayon) [inline llvm/lib/Target/EraVM/EraVMInstrFormats.td]: Could you elaborate? I am not sure I understood what is going on here.

File: llvm/lib/Target/EraVM/EraVMInstrInfo.td
Comment (sayon) [inline llvm/lib/Target/EraVM/EraVMInstrInfo.td:131]: Question (not for this PR): can/should we get more operand types, for code and stack in different addressing modes?

File: llvm/lib/Target/EraVM/EraVMInstrInfo.td
Comment (sayon) [inline llvm/lib/Target/EraVM/EraVMInstrInfo.td:1094]: This instruction is marked with `hasSideEffects` in the resulting tablegen, but it is basically outputting a constant value within a single contract execution instance. It can be cached, it can be read at any place and freely moved around etc.

File: llvm/lib/Target/EraVM/EraVMInstrInfo.td
Comment (sayon) [inline llvm/lib/Target/EraVM/EraVMInstrInfo.td:1081]: I wonder if this instruction should be marked as `hasSideEffects`. The only constraint that we need to enforce is that SetU128 would be executed before the next far call in the trace.

File: llvm/lib/Target/EraVM/MCTargetDesc/EraVMAsmBackend.cpp
Comment (sayon) [inline llvm/lib/Target/EraVM/MCTargetDesc/EraVMAsmBackend.cpp]: IMO we could use more types here nicely to get rid of the (situationally used) arguments. 
Like a struct and an overload of `appendMCOperands` for each argument type.

File: llvm/lib/Target/EraVM/EraVMInstrInfo.cpp
Comment (sayon) [inline llvm/lib/Target/EraVM/EraVMInstrInfo.cpp:902]: Why do we omit the CTXr_se (did it become "gasleft" or was it removed?)

File: llvm/test/CodeGen/EraVM/intrinsic.ll
Comment (sayon) [inline llvm/test/CodeGen/EraVM/intrinsic.ll:21]: Why do we need to write the fullest form `[0 + reg]` here and in other tests?

File: llvm/test/CodeGen/EraVM/intrinsic.ll
Comment (sayon) [inline llvm/test/CodeGen/EraVM/intrinsic.ll:384]: Should we treat these in the same way as `gasleftr`?

File: llvm/test/MC/EraVM/asm-parser/arith-instructions-errors.s
Comment (sayon) [inline llvm/test/MC/EraVM/asm-parser/arith-instructions-errors.s:30]: Some additions:
- arith instructions can not have any other modifiers (like is first, heap variant selectors etc.)
- repetition: `add.s.s.s` should fail

File: llvm/test/MC/EraVM/asm-parser/arith-operands-errors.s
Comment (sayon) [inline llvm/test/MC/EraVM/asm-parser/arith-operands-errors.s:23]: also `stack + [anything]`

File: llvm/test/MC/EraVM/asm-parser/events.s
Comment (sayon) [inline llvm/test/MC/EraVM/asm-parser/events.s]: Syntax should be coherent now, from what I see. Please ping if the problem persists.

File: llvm/test/MC/EraVM/asm-parser/sp-changes.s
Comment (sayon) [inline llvm/test/MC/EraVM/asm-parser/sp-changes.s:14]: There is no test for `nop` without any operands.

File: llvm/lib/Target/EraVM/EraVMInstrInfo.td
Comment (asl) [reply llvm/lib/Target/EraVM/EraVMInstrInfo.td:131]: For this you'd need to split the instructions. So, you'd end with # of instruction blow-up.

File: llvm/lib/Target/EraVM/EraVMInstrInfo.cpp
Comment (asl) [reply llvm/lib/Target/EraVM/EraVMInstrInfo.cpp:902]: As I mentioned: we created separate instruction instead of "contex operand". So, there is no `CTXr_se` and similar instructions anymore (see .td file).

File: llvm/test/CodeGen/EraVM/intrinsic.ll
Comment (asl) [reply llvm/test/CodeGen/EraVM/intrinsic.ll:21]: This is just how the things are emitted for now. We have not added simpler "forms".

File: llvm/test/MC/EraVM/asm-parser/arith-instructions-errors.s
Comment (asl) [reply llvm/test/MC/EraVM/asm-parser/arith-instructions-errors.s:30]: @atrosinenko Please add cases

File: llvm/lib/Target/EraVM/EraVMInstrInfo.td
Comment (asl) [reply llvm/lib/Target/EraVM/EraVMInstrInfo.td:1094]: Yep. But this is how it was previously (it was `CTXr_se` instruction that had `hasSideEffects = 1` set).

File: llvm/lib/Target/EraVM/EraVMInstrInfo.td
Comment (asl) [reply llvm/lib/Target/EraVM/EraVMInstrInfo.td:1081]: Well, I believe we kept the pre-existing behavior. So, if some things could be relaxed, this could be done in a separate change.

File: llvm/test/CodeGen/EraVM/intrinsic.ll
Comment (sayon) [reply llvm/test/CodeGen/EraVM/intrinsic.ll:21]: thanks for clarifying!

File: llvm/lib/Target/EraVM/EraVMInstrInfo.td
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EraVM/EraVMInstrInfo.td]: In some optimizations, we utilize the implicit Flags register to identify the defs and uses of it. With this change, we would always assume that `JCs` instruction is using flags, which could be incorrect (as it is shown in the test). Even thought I don't think this change currently creates an issue, my suggestion is to keep this instruction, and think about how to remove this overhead later on. Wdyt?

File: llvm/lib/Target/EraVM/EraVMOpcodes.td
Comment (atrosinenko) [reply llvm/lib/Target/EraVM/EraVMOpcodes.td]: Fixed

File: llvm/lib/Target/EraVM/EraVMExpandPseudoInsts.cpp
Comment (atrosinenko) [reply llvm/lib/Target/EraVM/EraVMExpandPseudoInsts.cpp]: Grepped for `addImm(0)`, replaced with `addImm(EraVMCC::COND_NONE)` where appropriate.

File: llvm/test/MC/EraVM/asm-parser/arith-instructions-errors.s
Comment (atrosinenko) [reply llvm/test/MC/EraVM/asm-parser/arith-instructions-errors.s:30]: Fixed

File: llvm/test/MC/EraVM/asm-parser/arith-operands-errors.s
Comment (atrosinenko) [reply llvm/test/MC/EraVM/asm-parser/arith-operands-errors.s:23]: Added two test cases:
```
  add  stack+[@global + 1 + r1], r2, r3
  add  r1, r2, stack+[@global + 1 + r3]
; CHECK:       <stdin>:{{[0-9]+}}:14: error: '=' expected
; CHECK-NEXT:    add  stack+[@global + 1 + r1], r2, r3
; CHECK-NEXT:               ^
; CHECK:       <stdin>:{{[0-9]+}}:22: error: '=' expected
; CHECK-NEXT:    add  r1, r2, stack+[@global + 1 + r3]
; CHECK-NEXT:                       ^
```

File: llvm/test/MC/EraVM/asm-parser/events.s
Comment (atrosinenko) [reply llvm/test/MC/EraVM/asm-parser/events.s]: Removed `FIXME`

File: llvm/test/MC/EraVM/asm-parser/sp-changes.s
Comment (atrosinenko) [reply llvm/test/MC/EraVM/asm-parser/sp-changes.s:14]: Added `NOPrrr` instruction and `nop r0, r0, r0 -> nop` alias + test cases.

File: llvm/lib/Target/EraVM/EraVMInstrFormats.td
Comment (atrosinenko) [reply llvm/lib/Target/EraVM/EraVMInstrFormats.td]: Fixed

File: llvm/lib/Target/EraVM/EraVMInstrInfo.td
Comment (atrosinenko) [reply llvm/lib/Target/EraVM/EraVMInstrInfo.td]: Fixed: `s/to make disassembler parse/to make asm parser handle/` (3 instances).

File: llvm/lib/Target/EraVM/EraVMInstrInfo.td
Comment (atrosinenko) [reply llvm/lib/Target/EraVM/EraVMInstrInfo.td]: Dropped the commit that removes `J_s` for now.

File: llvm/lib/Target/EraVM/EraVMInstrFormats.td
Comment (asl) [reply llvm/lib/Target/EraVM/EraVMInstrFormats.td:105]: I've created a separate issue for this. Let's do afterwards.

**Comment by asl** [comment]
@hedgar2017 @akiramenai @antonbaliasnikov regression failures look like CI issue. Will you please check what is going on? Thanks!

**Comment by asl** [comment]
@antonbaliasnikov I think https://github.com/matter-labs/era-compiler-llvm/commit/6e708f8c2bde3b20eb2186e9ec00ce9acf0bad33#diff-f0de00012d74c09571bcabcb38ab9ba495177a8e76c00e5ab27cc887118e9b79R62 does not look properly with auto-generated files (during the build process) and LLVM has plenty of them.

File: llvm/lib/Target/EraVM/EraVMInstrInfo.td
Comment (atrosinenko) [reply llvm/lib/Target/EraVM/EraVMInstrInfo.td]: Finally had to introduce a separate commit https://github.com/matter-labs/era-compiler-llvm/pull/448/commits/fbdba35e29de815a70f550c76e85aded75635b2b and fold the expansion into https://github.com/matter-labs/era-compiler-llvm/pull/448/commits/3090f94936e71b1ea18274d1fa73419ce6e9f674, so J_s is properly expanded at any point in history (though, it seems integration tests are only executed for the HEAD commit of the branch).

**Comment by hedgar2017** [comment]
@asl this can happen if the branch is not rebased. Benchmarks and run against a newer `main` which is more efficient.
Rebasing now.

**Comment by asl** [comment]
> Benchmarks and run against a newer `main` which is more efficient.
> Rebasing now.

I do not think this is the issue. As these are per-commit regression tests failing due to being unable to build LLVM. We checked locally these revisions and they are fine. So, definitely a CI issue.

**Comment by hedgar2017** [comment]
@asl understood, I thought the slight changes in benchmark were the problem, but I'm OK with ignoring them.
@antonbaliasnikov could you check why `benchmark` succeeds and `default` and fails? Please fix in `main` and we'll rebase.

**Comment by sayon** [comment]
I can only build the latest commits in debug mode, otherwise I am getting errors like that one:

```
Undefined symbols for architecture arm64:
  "llvm::MachineInstr::dump() const", referenced from:
      (anonymous namespace)::EraVMAsmPrinter::emitInstruction(llvm::MachineInstr const*) in libLLVMEraVMCodeGen.a[5](EraVMAsmPrinter.cpp.o)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
[199/296] Linking CXX executable bin/bugpoint
```

(thanks to @vladimirradosavljevic)

**Comment by asl** [comment]
@sayon Are you running a build w/o asserts? Which revision fails in such way?

**Comment by sayon** [comment]
@asl yes without asserts (for building multiple artifacts, the debug build is sometimes painfully slow).

I used commit 3ecd9c4b40e585bd6c7d79d14352d1d13ac5f2d8.

**Comment by asl** [comment]
> @asl yes without asserts (for building multiple artifacts, the debug build is sometimes painfully slow).
> 
> I used commit [3ecd9c4](https://github.com/matter-labs/era-compiler-llvm/commit/3ecd9c4b40e585bd6c7d79d14352d1d13ac5f2d8).

Thanks. This might be simple to fix.

**Comment by asl** [comment]
> @asl yes without asserts (for building multiple artifacts, the debug build is sometimes painfully slow).

Ok, this was fixed.

File: llvm/test/MC/EraVM/asm-parser/sp-changes.s
Comment (sayon) [reply llvm/test/MC/EraVM/asm-parser/sp-changes.s:14]: I have double checked with the existing assembler and the syntax `nop stack+=[reg+imm]` as well as `nop stack -= [reg + imm]` are accepted; I have updated the spec -- please add the test for these cases too.

**Review by sayon — CHANGES_REQUESTED** [review]

**Comment by atrosinenko** [comment]
I pushed an initial version of the branch rebased to LLVM 17 to run CI as early as posible. Though, more changes have to be done, at least the very first commit has to be folded somewhere.

**Comment by atrosinenko** [comment]
* dropped a few changes - [reverted](https://github.com/matter-labs/era-compiler-llvm/compare/ac0e8c554a032602ffe6826dfc21b0ae0bc22a78..e0a87253a6bf59fa2a994a419d16c7121e18544a) back to overloaded constructors in `EraVMOperand` as there are only four now (exactly one per union member), so passing lambdas to static function is too complicated
* dropped a temporary commit as #486 is merged now

**Review by sayon — APPROVED** [review]

**Review by akiramenai — APPROVED** [review]
LGTM, thanks.
I'm going to wait until the integration tests pass and then merge this PR.

**Comment by antonbaliasnikov** [comment]
> LGTM, thanks. I'm going to wait until the integration tests pass and then merge this PR.

@akiramenai , there are some problems with our runners who cannot pick up jobs. I canceled them for now, but I'll restart them as soon as the problem is fixed and let you know.

**Comment by antonbaliasnikov** [comment]
@akiramenai , please, rebase this branch from the latest `main` to take CI fixes. CI should work now.

**Comment by asl** [comment]
@akiramenai Looks like there is unresolved conversation that was attached to a commit that does not exist anymore. As a result, the PR cannot be merged :) For the record, the comment was resolved.

**Pull request merged**
